### PR TITLE
Braid works with looparound and negative indices

### PR DESCRIPTION
**Pull request by SzilBalazs**

### CODE REVIEW

**Comment by sulan** [comment]
Please rebase against main

File: knpy/braid.py [hunk 1]
```diff
@@ -122,18 +122,17 @@ def braid_relation1(self,index,inplace=True):
...
+    def braid_relation2(self,index):
         """
         Perform second braid relation.
         index: Where the chunk starts, on which operation can be done
         """
         if self.is_braid_relation2_performable(index):
+            if index >= 0 and index < len(self._braid):
+                index -= len(self._braid)
             transformed_braid = self._braid
             transformed_braid[index], transformed_braid[index+1] = transformed_braid[index+1], transformed_braid[index]
```
Comment (sulan) [inline knpy/braid.py:134]: This line modifies `self._braid`.

File: knpy/braid.py [hunk 2]
```diff
@@ -122,18 +122,17 @@ def braid_relation1(self,index,inplace=True):
...
 
-    def braid_relation2(self,index,inplace=True):
+    def braid_relation2(self,index):
         """
         Perform second braid relation.
         index: Where the chunk starts, on which operation can be done
         """
         if self.is_braid_relation2_performable(index):
+            if index >= 0 and index < len(self._braid):
+                index -= len(self._braid)
```
Comment (sulan) [inline knpy/braid.py:132]: Why?

File: knpy/braid.py [hunk 3]
```diff
@@ -237,10 +236,15 @@ def braid_relation1_performable_indices(self):
         return np.array(positions)
 
     def is_braid_relation2_performable(self,index):
-        return self._braid.shape[0] != 0 and 0 <= index and index + 1 < self._braid.shape[0] and abs(abs(self._braid[index]) - abs(self._braid[index+1])) >= 2
+        if index >= 0:
+            index -= len(self._braid)
```
Comment (sulan) [inline knpy/braid.py:240]: Why?

File: test/test_braid.py [hunk 4]
```diff
@@ -329,31 +329,43 @@ def test_braid_relation2_empty(self):
...
     def test_braid_relation2_negative_exception(self):
         braid = Braid([1, -3, 2])
-        for i in range(1, 5):
+        assert braid.braid_relation2(-3) == Braid([-3, 1, 2])
+        for i in range(-2, 0):
             with pytest.raises(IllegalTransformationException):
-                braid.braid_relation2(-i)
-            
+                braid.braid_relation2(i)
+        for i in range(-5, -3):
```
Comment (sulan) [inline test/test_braid.py:356]: What exactly is the range of the allowable indices? There is only the check for `0 <= index <= len`, so what happens in the other "periods"? (If we want to allow such indices.)

**Review by sulan — CHANGES_REQUESTED** [review]
And please rebase

File: knpy/braid.py (hunk 2 above)
Comment (Error-42) [reply knpy/braid.py:113]: Comment, documentation added; exceptions handled more clearly.

File: knpy/braid.py (hunk 3 above)
Comment (Error-42) [reply knpy/braid.py:216]: Comment added; exceptions handled more clearly.

File: test/test_braid.py (hunk 4 above)
Comment (Error-42) [reply test/test_braid.py:343]: See updated docs of braid_relation2; exceptions handled more clearly.

**Review by sulan — APPROVED** [review]

**Pull request merged**